### ISSUE TITLE
Add a system of page-less nodes

### ISSUE THREAD

**Issue by saltednut** [backdrop/backdrop-issues#100]
Since we will soon have relegated blocks to being an admin-only text-only layout element that's saved into config, we are going to need a content-based alternative that is a fieldable entity!

Big picture, what we want are page-less nodes. 

We have several examples of entities that come close in Drupal contrib that we can compare/contrast and decide if we want to start from any of them, or would prefer to create our own.

---
Use Cases
---

1) Fieldable and typable block content:
I often need to create a type of content that will only be viewed as a Block. 
**Example**: an `Ad block` that consists of an image field, a link field, an (optional) text field, and can be placed into a layout. 
**Solutions**: [Bean](https://www.drupal.org/project/bean) or [entity construction kit](https://www.drupal.org/project/eck)

2) Content without a "page"
I often need to create a type of content that will only be seen in a view. 
**Example**: a "Slider item" that consists of an image field, a text field, an optional description field and an optional link field. 
**Solutions**: [entity construction kit](https://www.drupal.org/project/eck) or [Rabbit Hole module](https://www.drupal.org/project/rabbit_hole)

3) Content as part of other content
I often need to create a type of content that will only be seen as part of another piece of content. 
**Example**: Example: media-rich content. Sometimes content editors want to insert some media items between text paragraphs: sliders, photo galleries, maps, video, etc.
**Solution**: [paragraphs](https://www.drupal.org/project/rabbit_hole) or [entity construction kit](https://www.drupal.org/project/eck) + [inline entity form](https://www.drupal.org/project/inline_entity_form) or [field collection](https://www.drupal.org/project/field_collection) or  [multifield](https://www.drupal.org/project/multifield)

~~4) Grouping fields together (**not a great fit for pageless nodes**)
I often need to group my fields together, so that when I add another "field" I'm actually adding another "Set of X fields". 
**Example**: A "contact person" may consist of a name field, email field, and phone field.
**Solutions**: [field collection](https://www.drupal.org/project/field_collection) or  [multifield](https://www.drupal.org/project/multifield)~~

----
PR by @serundeputy: https://github.com/backdrop/backdrop/pull/2049
New PR, rerolled and tweaked by @herbdool: https://github.com/backdrop/backdrop/pull/2286

**Comment by jenlampton** [comment]
Wow, it's been a while since we looked at this issue. I, for one, would love blocks to be entities and for them to be fieldable in core.  However, I don't think this can a happen in 1.x anymore.

**Comment by MrHaroldA** [comment]
What if ...

1. We add another optional view mode to nodes: `Block`
2. We add another checkbox to the publishig options for content types that have the Block view mode active: `[ ] Expose as block`
3. Nodes with the checkbox show up as block

Is this "oversimpling" the issue?

**Comment by MrHaroldA** [comment]
Here's a proof-of-concept that exposed a node as a block, if checked in the publishing options.

https://github.com/backdrop/backdrop/compare/1.x...MrHaroldA:block_nodes?expand=1

It lacks a lot of things; but it's basically a "fieldable block system" of some kind ... ;)

**Comment by MrHaroldA** [comment]
It would be cool if someone could check the architectural changes of my patch, and tell me if it's viable or not.

I'd expect it to be difficult to achieve, but my patch is really straight-forward ...

**Comment by mikemccaffrey** [comment]
> @jenlampton: I, for one, would love blocks to be entities and for them to be fieldable in core.

I've been thinking a lot about this, and I have not been able to come up with many cases where it would make sense to add fields to blocks. Pretty much the only one would be a field to hold images that are used in the block.

Do folks have other examples where'd they add fields to blocks? And would these fields just be added to all the blocks on the system, or do we have to implement the concept of block types?

**Comment by mikemccaffrey** [comment]
A reply from @jenlampton with examples of fieldable blocks that people might create: 

> ex: type: slide - a home pageslide that needs fields (video, image, text, link) and is only accessed via a view
> ex: type: ad - an Ad needs fields (image, link) but is placed individually in a layout
> ex: type: administrator - this item appears only within various lists of administrators. This content will not be associated with a user account, but needs fields (name, address, phone number) as well as categorization.

This will obviously not only require us to convert to the existing custom blocks to entities, but also provide support for different types of blocks.

**Comment by MrHaroldA** [comment]
What about my "node as block" idea?

**Comment by mikemccaffrey** [comment]
I see where @jenlampton is coming from with the fieldable block types above, as I've created my share of slide and ad content types that didn't necessarily need to have their own public pages and all the bells and whistles that come with nodes, such as author and comments.

However, I very rarely needed to place these elements directly into the page as individual blocks, and instead usually have a view that displays either a list of these items or a single random item. In fact, it would be kinda overwhelming if a new block was created for every slide or ad that is added over the life of the site.

Therefore, while I'm on board with the idea of creating a new entity for creating fieldable types of content that don't need their own page, do we really want to call them blocks?

**Comment by mikemccaffrey** [comment]
> @MrHaroldA: What about my "node as block" idea?

I like it this idea. Perhaps, just like we are creating a block that allows you to place individual fields into a layout in #1100, we could create another block type that allows you to place a whole node into the region, which has nid and display mode as settings. That might be simpler than defining a unique block for each node that you want to be able to use in a layout.

**Comment by klonos** [comment]
> What about my "node as block" idea?

@MrHaroldA: do you have an issue filed for this already? Because the closest we have to that AFAIK is #1076.

**Comment by MrHaroldA** [comment]
Not really, but I did make a proof of concept right here: https://github.com/backdrop/backdrop/compare/1.x...MrHaroldA:block_nodes?expand=1

**Comment by mikemccaffrey** [comment]
Hey folks, I've been thinking about the idea of fieldable blocks, and it seems like the big requirements we have been asking for are:
- they are an entity
- they don't have all the extras that are baked into nodes
- you can create different types
- you can add arbitrary fields
- you can list them using views
- you can add them individually to layouts

However, based on that, it seems like we already have a drupal entity that does all of those things (except for the backdrop-specific layout integration), and that is field collections.

For those who are asking for fieldable blocks (like @jenlampton), are there any features that you are looking for that are not supported by field collections? What if we were to take the existing implementation of field collections, ported it to backdrop, and perhaps changed the name to describe how you'd like to use them?

**Comment by MrHaroldA** [comment]
Field collections have way more features than we need, is painfully slow on nested collections, and has (or used to have?) big problems with revisions. I also dislike the default edit and delete links on the display within a node, but those can be disabled.

For example: each field collection type has it's own page callbacks, including view, edit and delete.

**Comment by jenlampton** [comment]
I'm also in agreement about field collection being bloated and (more importantly) really slow. I'd love to see something architecturally cleaner than that, something more like a page-less node.

**Comment by serundeputy** [comment]
On the last project I worked on we made extensive use of the [paragraphs module](https://www.drupal.org/project/paragraphs).

Which is basically a fieldable entity that is not a node.  You can attach a paragraph bundle to a node or use it in a view, but it is not a block.

Just throwing that out there.  I don't exactly understand all the points of view here (that doesn't make them bad), but paragraphs was a nice solution for me for some of the things discussed here.

\#2¢geoff.

**Comment by mikemccaffrey** [comment]
> @serundeputy: On the last project I worked on we made extensive use of the paragraphs module.

Good call. That project seems very pertinent to what we are looking for, and we should examine the architecture of that module closely, even if we don't do a direct port or end up building things in the exact same way.

**Comment by darol100** [comment]
Another possible solution for this problem will be to port [Rabbit Hole module](https://www.drupal.org/project/rabbit_hole). Here is the description of Rabbit Hole. 

> Rabbit Hole is a module that adds the ability to control what should happen when an entity is being viewed at its own page.

>Perhaps you have a content type that never should be displayed on its own page, like an image content type that's displayed in a carousel. Rabbit Hole can prevent this node from being accessible on its own page, through node/xxx.

 Instead of trying to port a complex module like ECK, we can try to port Rabbit Hole which prevents from pages from been seen at the node level.

**Comment by MrHaroldA** [comment]
@darol100 It seems like a waste to me to throw away things php has built in the same process. Disabling detail pages on a content type should prevent them from being built, not thrown away afterwards.

**Comment by mikemccaffrey** [comment]
We mentioned this issue again in the call last week, and I brought up this question: how are these new page-less entities different than nodes other than the lack of a URL? What would the base table look like? Would it have the same attributes as nodes (author, created, updated, etc.)?

I feel like knowing that would help us understand if these page-less nodes are really a different species of creature.

**Comment by klonos** [comment]
> We mentioned this issue again in the call last week, and I brought up this question: how are these new page-less entities different than nodes other than the lack of a URL?

When you mentioned this, I realized that I had the same question in my head before, but couldn't exactly find the right words to phrase it. As a non-coder, I did not have the db table concerns etc, but honestly, the page-less node thing in my head could be accomplished if we somehow allowed and left the path empty (or with a special value like `null` or `<none>`). The technical implementation is of course beyond me at this point.

**Comment by jenlampton** [comment]
Pushing to 1.7 since there has been no activity on this issue and code-freeze is in just over 3 weeks.

**Comment by Al-Rozhkov** [comment]
There is one module from D7 which wasn’t mentioned in this issue. It is [Fieldable Panels Panes](https://www.drupal.org/project/fieldable_panels_panes). It has one important feature, it is possible to add pane directly from Panels UI.
> how are these new page-less entities different than nodes other than the lack of a URL?

So, I think it is the main difference for page-less entities from regular nodes in UX sense. Because it is very uncomfortable to create node then insert it in proper layout. Much better UX for me is to create fieldable blocks directly in Layout UI (like custom blocks) with ability to make them reusable.

Field collection or Paragraphs is another way to build composite content. Paragraphs items tied to host entity, at least it is the only way to create them. People use Paragraphs to build complex pages and even long landing pages. With power of fields and field formatters it is easy to give end-user ability to create map or gallery inside single node. Modules like [Classy paragraphs](https://www.drupal.org/project/classy_paragraphs) exists to style Paragraphs items. But Backdrop already has Layouts and styles for blocks. So as site builder I dream only about fieldable block types in Backdrop.

By the way Paragraphs extremely slow too. And editorial workflow becomes very painful when you need to edit one small thing inside big list of items.

**Comment by jenlampton** [comment]
This is too big an issue go get done in one day with no PR yet, unless @serundeputy has a trick up his sleeve, I'm bumping to 1.8.

**Comment by Al-Rozhkov** [comment]
I wish I can help with that, but I don't understand what direction was chosen by the community.

- Introduce new entity type.
- Add setting to node types.
- Other :smile:

**Comment by jenlampton** [comment]
>  I don't understand what direction was chosen by the community.

This is the crux of the problem. We haven't decided. I think in general we are leaning towards less abstraction, and potentially all "content" being nodes, but @mikemccaffrey brought up some good questions that we need to answer before we can make the call.

I created a little spreadsheet of the entities we currently have in core, their associated medatata, and the features those map to.
https://docs.google.com/spreadsheets/d/1rY1XM5_mxyCnISaY7kny1aRPGJ3q5AXyv-SGNTtVLD4/edit#gid=0

In some cases, we currently have entities that are missing features we probably need (like translatable and revisionable terms, for example). So we should think about this question at a higher level than just comparing features on nodes and blocks.

**Comment by mikemccaffrey** [comment]
I think that this discussion has been a bit muddled since we started it by declaring a potential solution and diving into the technical details without fully describing the problem that we are actually trying to solve. And based on people's comments, it seems like there might be several different use cases that people are looking to implement with "pageless nodes".

Perhaps folks could submit some other tickets that better document the user needs that brought them to this ticket? What exactly are you trying to accomplish? How are you currently doing it? Then we can discuss the different ways that can be accomplished, with these pageless nodes perhaps being only of the options.

Make sense?

**Comment by jenlampton** [comment]
I added 4 use cases for pageless nodes to the first issue in this thread based on the modules people recommended as solutions to this problem. (I don't actually think #4 is a good use-case for pageless nodes, but I added it there anyway for reference).

I haven't used Paragraphs yet, so I figured I should leave a real world example for that one to someone else.

Are there others?

**Comment by Dinornis** [comment]
I used num.2 example on D6 site.
Needed to use content without a "page" to create a type of content only accessible in a view.

**Comment by Al-Rozhkov** [comment]
The most common use case for my sites is number 3. **Example:** media-rich content. Sometimes content editors want to insert some media items between text paragraphs: sliders, photo galleries, maps, video, etc. 

![paragraphs_abstract_0](https://cloud.githubusercontent.com/assets/16973264/25749274/0550a72e-31bf-11e7-9c56-e3b647ebaf2e.jpg)

Picture taken from D7 paragraphs module.

But at the same time I need Fieldable and typable blocks. Actually, I have a site with Paragraphs and Fiedable Panels Panes modules installed together with very similar bundles and fields. Paragraphs for media-rich content inside nodes and FPP for blocks. I feel this task can be done more elegantly without duplication.

So my vote for number 1 and 3 :-)

**Comment by mikemccaffrey** [comment]
> I added 4 use cases for pageless nodes to the first issue in this thread based on the modules people recommended as solutions to this problem.

So, basically, it seems like we have four different problems that have been lumped into a single ticket. Based on all the back and forth, it doesn't seem like we can come up with a single solution to meet the needs of all of the different use cases, so perhaps we should create four different issues to discuss how to independently support each of them in core?

**Comment by jenlampton** [comment]
> So, basically, it seems like we have four different problems that have been lumped into a single ticket. 

Yes, that's because they all have something in common and it could be that there's a single solution for all of them.

> it doesn't seem like we can come up with a single solution to meet the needs of all of the different use cases

I'm not sure how you reached this conclusion already. I would hate to see four separate implementations of pageless nodes in core when it could be a single solution (a new entity type?), with 4 different user interfaces (create from layouts UI, create from content content screen, create as fields on nodes, and group togther as field groups).

> perhaps we should create four different issues to discuss how to independently support each of them in core?

I don't think this is a good idea... at least not until we know what the solution(s) will be.

**Comment by jenlampton** [comment]
For example, I would be happy to remove the field_collection / multifield use case since it is quite a bit different than all the others, and I think it's fine to leave that solution to contrib.

**Comment by jenlampton** [comment]
I also feel that Paragraphs is a great (contrib) solution for number 3, and I don't think we need an issue for including it in core either.

Item #2 could be solved easily by a rabbit-hole type implementation. But if we went that route for Item #1, we would also need an interface for creating this new type of content from the Layout UI, like we do with blocks.

By looking at only 2 use-cases, the rabbit-hole approach looks more straightforward. However, if we look at all 4, maybe a new entity type appears more attractive? (Paragraphs, Field collection, and Multifield all create a new entity type).  Or, maybe we decide that new entity types should be left to contrib, and in core we do only what's simplest?

I'm in favor doing what's simplest. :) 

How about adding a setting to all node types that would allow that type of content to be created directly from the Layouts UI (or not).  If we added a parallel (rabbit_hole) setting that allowed that type content to be accessed via a specific URL (or not) It could be possible to get both these changes in during the 1.x cycle. :)

**Comment by olafgrabienski** [comment]
> How about adding a setting to all node types that would allow that type of content to be created directly from the Layouts UI (or not). If we added a parallel (rabbit_hole) setting that allowed that type content to be accessed via a specific URL (or not) It could be possible to get both these changes in during the 1.x cycle.

Sounds good to me!

I'd love to use *Paragraphs* as a contrib module as well. A port started a while ago, haven't tested it so far. Here the project links:

- https://github.com/backdrop-ops/contrib/issues/212
- https://github.com/jacksinn/backdrop-paragraphs

Apart from that, I guess even the rabbit-hole approach would allow a simple implementation of use-case number 3, if used in combination with *Reference(s)*.

**Comment by LeeteqXV** [comment]
https://github.com/backdrop-ops/forum.backdropcms.org/issues/22

**Comment by jenlampton** [comment]
Hm, we still don't have any code for review here and we're 2 weeks from feature freeze. I think this is too big a feature to get into the next release in the time we have, so removing from 1.8 milestone. We can add a milestone back as soon as we have code to review.

**Comment by jlfranklin** [comment]
We have a handful of great modules, all of which solve the problem in very different ways.  Can we start by porting them to Backdrop and let site builders vote through module popularity?

**Comment by jenlampton** [comment]
Yes, that's a good way to help us decide :)

**Comment by olafgrabienski** [comment]
> We have a handful of great modules, all of which solve the problem in very different ways. Can we start by porting them to Backdrop and let site builders vote through module popularity?

As far as I know, respective modules were not ported so far, or are not ready. After re-reading the issue I have the impression that the Rabbit Hole approach mentioned lately in https://github.com/backdrop/backdrop-issues/issues/100#issuecomment-299574518 would solve most use cases. @laryn ported the module a while ago but it's not ready to use. Is anyone up to help there? https://github.com/laryn/rabbit_hole (Or should we even still port / adapt Rabbit Hole directly in core?)

**Comment by jenlampton** [comment]
I'd like to help with the port of Rabbit hole, and I think it would be good to get it working for contrib before we put it into core. We'll likely need to adapt it for core, but having a working contrib version is a useful first step!

**Comment by herbdool** [comment]
@olafgrabienski @jenlampton @laryn the basics are working and needs testing. Need to figure out why it's not for taxonomy yet.

**Comment by olafgrabienski** [comment]
@herbdool, great news! Will test as soon as possible, hopefully tomorrow.

**Comment by olafgrabienski** [comment]
@herbdool Have tested Rabbit hole now, looks great! Found only minor issues, see my comments on the existing page title issue and two new issues: https://github.com/backdrop-contrib/rabbit_hole/issues

**Comment by serundeputy** [comment]
I took the ideas from rabbit hole and implemented a simplified node only option to `Hide nodes`.
@olafgrabienski this will allow us to make a content type for translatable blocks, so we sort have a fieldable content block; and since it is a node we can use the node translation system to translate these blocks as well.

QC Env:
Website: http://2049.backdrop.backdrop.qa.backdropcms.org
Username: admin
Password: 7TZqGNWd

**Comment by olafgrabienski** [comment]
@serundeputy Nice! Have already built a respective content type in the environment (and found myself on a not found page directly after creating the node). Will have a closer look in the next days.

**Comment by serundeputy** [comment]
Thanks @olafgrabienski !

I've pushed up a fix so that new content types can pick up a default of the usual behaviour of an accessible page.

**Comment by olafgrabienski** [comment]
To give an update: the new simplified option to 'hide' nodes (PR by @serundeputy) has been intensively discussed on the [page of the PR](https://github.com/backdrop/backdrop/pull/2049) itself. The user interface and the functionality of these *Pageless settings* (how they're called now) have been improved subsequently, so that we now have a good and simple approach for pageless nodes provided by Backdrop core. As far as I can tell, the PR works good but I guess it could need another review.

Supposing a positive review, there might still come up the question why we don't just use the recently ported [Rabbit Hole](https://github.com/backdrop-contrib/rabbit_hole). In my opinion, *Rabbit Hole* is fine, especially if you need more functionality (handling of not just nodes but also taxonomy terms and user accounts, or the option to overwrite content-type-wide settings individually).

In contrast, the core *Pageless settings* are only for nodes, which keeps it simple and fits exactly to the title of this issue ("Add a system of page-less nodes"). Another (potential) advantage of the core implementation would be better integration, for instance by adjusting related node options like *Menu settings* automatically when someone chooses to make a content type pageless. Finally, the *Pageless* option in core would facilitate to use a popular block translation approach, thus making Backdrop more multilingual out of the box.

That said, I'm not sure if the new *Pageless settings* should be enabled by default. It's another tab in the content type settings which might not be used very often, just guessing. What do you think, does it make more sense to provide the *Pageless settings* as a core module which is not enabled by default?

**Comment by herbdool** [comment]
@olafgrabienski while I agree it probably won't be a popular approach, we can't know for sure since so far this functionality was only offered by a contrib module in D7 (Rabbit Hole).

One reason to keep it as part of the node module is because it avoids the technical issue we encountered with RH where the page not found wasn't working fully.

**Comment by herbdool** [comment]
I was curious to test how people were thinking of using it for translated blocks. When I enable a second language and then translate a node and then insert the node into a layout directly, it doesn't show the translated node when switching languages.

I've used views myself to accomplish this but can't see how to do it with a single piece of content inserted into a layout.

**Comment by MrHaroldA** [comment]
I haven't touched Backdrop in quite a while due to a new job, but here's what I tested:

1. Create a 'memow' with path alias 'test'
2. Create a 'page' with path alias 'test'
3. "The alias is already in use."

I'd recon that the pageless 'memow' shouldn't claim a url alias.

**Comment by herbdool** [comment]
Good point @mrharolda. We may need to disable path aliases and menu links for a content type with this enabled.

**Comment by olafgrabienski** [comment]
> I was curious to test how people were thinking of using it for translated blocks. When I enable a second language and then translate a node and then insert the node into a layout directly, it doesn't show the translated node when switching languages.

Oh my, you're absolutely right! When I insert a node in a layout, I have to pick content of a certain node ID. As a translated node has a different ID than the 'original' one, I have to decide which one to include: the original or the translation.

Thanks a lot for testing this, I should have done it long ago! (In a way it's funny, that we didn't noticed this in the whole discussion about [Translating blocks over in the Backdrop4Good issue](https://github.com/Backdrop4Good/backdrop4good.org/issues/39). (Will report the problem also there.)

At the moment, I'm not sure what to do regarding the not working translation approach. On first sight, it doesn't seem to be crucial for this issue which actually isn't about translation but about pageless nodes. On the other hand it feels strange to introduce the pageless option knowing that one of its main use cases (Fieldable and typable block content) ends up with not translatable content. Opinions?

**Comment by herbdool** [comment]
@olafgrabienski even though the more direct approach doesn't work, you can still create a views block and put that into the layout. Then assign blocks with a taxonomy term. The view block can filter by the term. Or put a reference field on a content type and show those in the layout as a field. Both aren't as direct but might work for you case

**Comment by olafgrabienski** [comment]
@MrHaroldA @herbdool The URL alias is an interesting topic. Do we need it on pageless nodes? Not necessarily because main use cases as SEO, or URL readability for visitors, are not so useful here. But URL aliases don't really hurt in my opinion, and they might however help editors or administrators  with readability.

What happened to @MrHaroldA could however happen with any other content type, be it with or without pageless setting.

That said, I agree with disabling menu settings / links for pageless content.

**Comment by olafgrabienski** [comment]
@herbdool Thanks for the views block idea. Do you intend a views block with Filter criteria "Current user's language"? Just an idea: Such a criteria could be a new approach to make node blocks language aware.

The Views Filter criteria "Language" includes:

- Current user's language
- Default site language
- No language
- English
- (other site languages)

In contrast, Block visibility conditions provide only:

- Default site language
- English
- (other site languages)

So, maybe we can add a Block visibility condition "Current user's language" to Backdrop and make sure that an existing content block recognizes the translation nodes. Just a first idea ...

**Comment by herbdool** [comment]
Yes, you need to use that filter in views.

It's a good idea for layout but it needs to know what to switch it to. It's got to be smarter about using the translation set on a node.

**Comment by mazzech** [comment]
Thank you @olafgrabienski for reporting... I am not sure if I understood all the points correctly, but IMO a "simple" additional language parameter for placing blocks in core could do the job, pretty similar to what Views does. This problem will affect any "inserted" content item in i18n projects, they have to be language aware (e.g. newsletter signups form). 

BTW, I hope to make it to Drupalcamp Ruhr, maybe we can deepen the discussion there:-)

**Comment by olafgrabienski** [comment]
@mazzech Thanks for your feedback! I think also that blocks could learn form Views. I'll try to check how Views works exactly in similar situations when I find time.

And yes, hope to see you at DrupalCamp Ruhr!

**Comment by olafgrabienski** [comment]
> you can still create a views block and put that into the layout. Then assign blocks with a taxonomy term. The view block can filter by the term.

@herbdool I'm not sure if I got your views block idea correctly. IIRC, as each block represents a single node, each block has to use a different term. Do you mean to filter by these terms with a contextual filter? If yes, I could need another hint how to configure it.

(When I assign all blocks with *one* taxonomy term and filter the view by this single term, all tagged blocks will be displayed in the layout.)

**Comment by herbdool** [comment]
@olafgrabienski I had to go back and look at how I did it. It was actually a bit rudimentary which was okay for the low complexity of the site. I created a block for each "snippet type" taxonomy term. The term just specified the location, e.g. front page button, blog page, etc. Then each translated node has the same term.

I'm sure there's a more dynamic way to do it with a contextual filter, off the top of my head I'm not sure.

**Comment by olafgrabienski** [comment]
@herbdool Thanks for your looking up your workaround, indeed reasonable for not too complex sites!

**Comment by olafgrabienski** [comment]
@ everyone interested in the nodes as blocks translation approach: I've finally filed a request to solve the current issue: [Blocks of existing content should follow the translation system](https://github.com/backdrop/backdrop-issues/issues/2957)

**Comment by alexfinnarn** [comment]
Hello all. I am very interested in this issue and would like to test out some of the proposed solutions. 

As I'm reading through this issue, it seems like the Rabbit Hole module is a viable option and the first thing to try? Or would you suggest other modules to test? 

For my use case, I'm trying to create "node layouts", essentially, that wrap around a node and provide a layout outside of the content section. The Layout admin UI is great, and I'd just like to have a "node/%/layout endpoint to control each node's version. 

Anyhwoo...the "pageless nodes" (we use Beans in D7) would be placed in these layouts so it would be nice to edit them from the Layout admin screen like was mentioned before.

I'd be happy to review the latest and greatest ideas in Backdrop core or contrib for implementing this feature.

**Comment by jenlampton** [comment]
hi @alexfinnarn and thanks for your interest in this topic. 

> As I'm reading through this issue, it seems like the Rabbit Hole module is a viable option and the first thing to try? Or would you suggest other modules to test?

Rabbit hole will prevent you from visiting the node/% page for any node type that you want to use as a block, in a layout. I'd definitely recommend giving it a try.

> For my use case, I'm trying to create "node layouts", essentially, that wrap around a node and provide a layout outside of the content section. 

Great, this is what Backdrop core provides when you create a layout at the path `node/%`. :)

> The Layout admin UI is great, and I'd just like to have a "node/%/layout endpoint to control each node's version.

What do you mean by `each node's version`? Are you looking for a different layout configuration for each node revision?

> the "pageless nodes" (we use Beans in D7) would be placed in these layouts so it would be nice to edit them from the Layout admin screen like was mentioned before.

We allow people to edit blocks from the Layout UI, so there's a precedent for allowing people to edit content from there too, but the "Existing content" block doesn't allow it, currently. 

Maybe this should become part of Rabbit Hole module, temporarily? We could provide a new block type that works much like "Existing content" (we could call it "New content", or something) and when used, it would allow people to create content of any type, and drop it into a Layout. Once placed, maybe we have an additional drop-button action for "Edit" that would allow editing of the content from the Layout UI, in addition to the "Configure" option.

**Comment by alexfinnarn** [comment]
The best way for me to explain what I'm trying to accomplish is probably to just show what we have in D7.

<img width="881" alt="screen shot 2018-07-17 at 11 06 01 am" src="https://user-images.githubusercontent.com/3640707/42826052-7c6a90ae-89b1-11e8-9ee6-57b418890d52.png">

Every node has a layout route that users can click on to alter what is in the sidebar, before/after content, and other regions defined in the active theme. The layout is active only for the current node and doesn't need a context like `node/%`...but maybe creating one layout for nodes would work, if I experimented more.

<img width="1313" alt="screen shot 2018-07-17 at 11 09 33 am" src="https://user-images.githubusercontent.com/3640707/42826235-eb388f36-89b1-11e8-83a7-74b57d644c35.png">

Once a user gets to the layout edit screen, they get a jacked up rendition of where the theme regions live and sections that pertain to adding existing or new blocks to the regions. I like the Backdrop forms and modal for editing/adding/rearranging blocks way better than what we have now. 

With the current layout system, I think a user would have to create a node, grab the node ID (3 for example), then make a layout with only `node/%` as the path and `node/3` visibility condition. I think I also would need to add the existing content around the other blocks in the layout.

We want to avoid all the jumping around and just tell users to click on "Layout", add their blocks, and be done with that individual node. We wanted to have base layouts that users could override, say a layout default for a content type that would be overridden when you saved a layout tied to that node.

All of this is ramble-y and spans many issues, but I was trying to make a contrib module I call "Node Layouts" to do what I mentioned above. So, this is just an example of a use case for "pageless nodes" or beans/fieldable blocks.  

I'll check out Rabbit Hole for your comments in the last paragraph, but we don't want some of the baggage of nodes, like authoring information. Since we know the Bean module, I guess we will be comparing options to that D7 module.

**Comment by klonos** [comment]
@alexfinnarn I believe that you are referring to #2894 and/or #1131 and/or #1126. Right?

I think that Rabbit Hole is a good fit for core. I was giving some media-related features in D8 a spin, and I saw that fieldable media assets by default get a `/media/MID` path (similar to `/node/NID`). If people want to hide them or even restrict access to them, they have very little options ATM (or they need to custom-code something). With a rabbit-hole-like functionality, this problem could easily be sorted in such cases.

**Comment by alexfinnarn** [comment]
For #2894, "What you suggest (different layouts per node) would be ideal..." and "Per-node layouts and per-content type layouts would make great contrib modules." Yes to that. The default layout is the content type layout and then when a user goes to "node/%/layout" they see how the blocks are placed in that layout and create a custom layout for that node if they save it with any changes.

For #1131, the user wouldn't be able to pick the layout but go to "node/%/layout" to arrange the blocks on a particular layout. We actually only really have one template so the user isn't choosing between those. It's just that every node can have different blocks in the layout, so I think there would be a layout per. I like the idea of choosing between predefined layouts with predefined blocks places in them, because that would reduce complexity and help with performance, but our users are used to the way it is now, and I don't think we can budge too much on what we setup. They seem to really like the option.

One idea would be to make a layout with the template we like and add a block to each region that dynamically places other content into it based on the current node. That way, out layout form can be saved and tied to a node ID. When the layout is rendered, the content of the block ties the "node/%/layout" values to the regions. I think that might work for our use case and would make it so only one layout would be stored in config rather than mix in content with the layout. I think it would also prevent layouts from pointing to deleted content/blocks...this might be my contrib module stab.

Thanks for pointing me to those issues @klonos! I will follow up in them instead of continuing in this issue since I am straying from the issue's main concerns/topic.

**Comment by docwilmot** [comment]
> Per-node layouts and per-content type layouts would make great contrib modules

I'd actually ported Panel Nodes a long while ago and forgot about it. A few issues to consider in the queue but it works.

**Comment by kevincrafts** [comment]
The issue I see with Rabbit Hole is that from the administration side, the content seems all the same and on a larger site I think site editors could get confused when they see node/page based content in the same list as node/block based content.

On a site that has 500 nodes and 300 blocks, not having distinct admin areas for them can complicate the work of the current editors and on boarding of new editors.

Keep in mind that @alexfinnarn are coming at this from the perspective that the large majority of our site editors are not web experts - they are doing all the changes on the site themselves, not contracting someone to do it for them.

**Comment by jenlampton** [comment]
> I'd actually ported Panel Nodes a long while ago

This is great!  It might also be worth looking at Panelizer. I suppose it depends on what editors were used to before. Because Layouts is a port of Panels, any panels-based module in D7 shouldn't be too hard to get working in Backdrop.

> On a site that has 500 nodes and 300 blocks, not having distinct admin areas for them can complicate the work of the current editors and on boarding of new editors.

As long as the "blocks" are nodes (and not blocks) separate admin areas for each will be a sinch using Views (or both Views and Layouts) :)

**Comment by herbdool** [comment]
This issue seems to cover a lot of different stuff but I think the meat is around a simple Rabbit Hole implementation (e.g. https://github.com/backdrop/backdrop-issues/issues/100#issuecomment-354512657). I've rerolled @serundeputy's PR and tweaked it https://github.com/backdrop/backdrop/pull/2286.

**Comment by olafgrabienski** [comment]
Thank you, @herbdool! I've tested the new PR on the sandbox site, it looks great!

**Quick report:**

- created a new content type, in *Display settings* enabled  the setting "Prevent visitors from seeing content on its own page"
- visited such content as Administrator => accessible = ok
- visited it as Anonymous => not found message = ok
- checked the default permissions => ok
- changed the permission for Anonymous => ok (after clear cache "Page and else")
- created a test page and placed a 'pageless' block of existing content => ok
- created a views block of 'pageless' content and placed it on the test page => ok
- created a respective views page => ok
- tested the settings and permissions on an existing content type (Post) => ok

**Issues:**

Found only one so far: After a "not found" message regarding 'pageless' content, I see also the following warning in the DBlog:
> Cannot modify header information - headers already sent by (output started at /home/qa/github/backdrop/backdrop/pr/2286/core/includes/bootstrap.inc:1793) in backdrop_send_headers() (line 1610 of /home/qa/github/backdrop/backdrop/pr/2286/core/includes/bootstrap.inc).

**Language related suggestions** (not a native English speaker, though):

(a) The term "pageless" is mentioned on the Permissions page but not in the Content type settings, I'd suggest to mention it for reference also in the Content type settings:

> Prevent visitors from seeing content on its own page.
> ~~The~~ **Pageless** content can still be visible in a view or a block, but will be hidden to those without the proper permissions when accessed as a full page.

(b) On the permissions page the use of the "pageless" term is a bit overwhelming, let's use one less:

Bypass pageless content restriction.
View pageless content ~~regardless of pageless setting~~ as a full page.

**Summary:**

- Looks very good.
- PHP warnings need work.
- Language questions may need feedback of a native speaker.
- We had already made many progress before. With this PR 'pageless' content could become part of 1.11 in my opinion.

**Potential follow-up ideas:**

- Disable 'pageless' setting for the Page content type delivered by core?
- Don't display menu settings on pageless content types.

**Comment by herbdool** [comment]
@olafgrabienski happy to improve the language. It's hard to explain the concept in just a few words so it can be improved.

I'm trying to figure out the conditions for when the error appears. I think it might be an issue of when the page is cached and then this module tries to override by showing a 404 page instead. E.g. error:

`Warning: Cannot modify header information - headers already sent by (output started at /home/qa/github/backdrop/backdrop/pr/2286/core/includes/bootstrap.inc:1793) in backdrop_serve_page_from_cache() (line 1766 of /home/qa/github/backdrop/backdrop/pr/2286/core/includes/bootstrap.inc).`

**Comment by herbdool** [comment]
@olafgrabienski I tweaked the code to try get rid of the "headers already sent" error. Can you give it another spin? I couldn't recreate the error consistently so just need to see if we can cause it.

**Comment by olafgrabienski** [comment]
> It's hard to explain the concept in just a few words

Indeed! So let's improve it a bit but not wait for the perfect solution.

> I tweaked the code to try get rid of the "headers already sent" error. Can you give it another spin? I couldn't recreate the error consistently so just need to see if we can cause it.

After your code tweak, I couldn't reproduce the error. To test it on the sandbox site, I visited a 'pageless' node as an anonymous visitor without the 'bypass' permission ("not found" as expected, no other error), then enabled the permission for the role so that anonymous visitors could visit the node without restrictions, and disabled the permission again (just "not found" again).

**Comment by olafgrabienski** [comment]
> when someone has permission to edit/create the content but does not have permission to view it on its own page. They would see "page not found" once they saved it.

Good point! As you say, "core redirects back to the front page". That happens for instance, when an editor has the permission to create or edit content, saves it as draft, and doesn't have the permission to view unpublished content. A comparable situation, in my opinion. Can we do the same redirect to the front page (instead of the "page not found" message) if someone has permission to edit/create content but not to view it on its own page?

Regarding the upcoming node preview, I'd say we shouldn't prevent a preview of the default / full display mode even if the editor doesn't have the pageless bypass permission. The preview is a part of the editing experience, and if an editor doesn't have the respective view permission, I can't imagine it intentional but due to a permission misconfiguration. (Same applies to the upcoming preview feature in case of a missing "view unpublished" permission, in my opinion.)

Re "combo of 'page not found and a permission to bypass" vs. "a permission to view the full display mode per content type", I find the latter more problematic at first sight: Wouldn't it mean, that a site builder had to enable this permission on all existing content types (except the pageless ones) for the anonymous and the authenticated role?

**Comment by herbdool** [comment]
@kevincrafts that's existing core functionality. As Olaf pointed out someone could have permission to create/edit but not have permission to "view unpublished". So that part isn't going to change. I'm pointing out the same situation applies with this PR (currently). A possible change (though I'm not sold on it yet) is to exempt authors (and anyone who can edit?) from the pageless restriction.

As for the preview, we'll have to see how it work once it's merged. I agree that they should be allowed to see a preview even if they can't view the full page.

At any rate, we're using the Rabbit Hole paradigm for this PR and thus has some oddities when being used with nodes which we built with paths.

**Comment by olafgrabienski** [comment]
> A possible change (though I'm not sold on it yet) is to exempt authors (and anyone who can edit?) from the pageless restriction.

Sounds reasonable!

@herbdool So far the issue has no milestone. Do you think it makes sense to tag it for 1.11, or would that be to hastily?

**Comment by quicksketch** [comment]
Well, we've got a few days, but the PR is much more limited. I'm down for trying, if we can get an implementation we're satisfied with. Thanks @herbdool for really simplifying this down to the main concept, I think it's a good direction.

**Comment by quicksketch** [comment]
I posted a code review https://github.com/backdrop/backdrop/pull/2286#pullrequestreview-151150847

> I'm also thinking that this combo of "page not found" and a permission to bypass is just more confusing than a permission to view the full display mode where it could be set up per content type and per role.

I agree, a permission to view the full node by type would be easier from an end-user perspective. However, I don't think we should attempt to cover that here. View permissions are not included in core because they can so easily get very complicated and sites have different requirements (hence the plethora of Node Access modules like Permissions by Term, Workbench Access, Domain Access, Organic Groups, [etc](https://www.drupal.org/node/270000)).

---

For reference, here's the current PRs language on enabling this option:

![image](https://user-images.githubusercontent.com/100206/44886490-3f6f1700-ac7b-11e8-9d09-a3dd80b87dbf.png)

- Regardless of the description on the checkbox (or radio button), we shouldn't end in a period for consistency.

Here's another stab at language:

- [ ] **Hide full page display**
If enabled, the URL for this content will return a "Page not found", preventing it from being accessed directly. It may still be placed as a block or within a listing. Users who have the "Access page-less content" will still be able to visit the full page URL.

To handle the "author may not be able to view the content" problem, we should modify the submit handler of the node form to redirect to the `node/x/edit` URL on save.

**Comment by herbdool** [comment]
Thanks for the code review @quicksketch. I've made some tweaks to the language. I used a slightly different permission: "View page-less content" since users can still edit and view as embedded so "access" didn't seem right.

I added the redirect to the edit page in the submit handler.

I didn't have any luck with using `MENU_NOT_FOUND` in that context. (See comment on review).

@olafgrabienski perhaps you want to take another look as well.

**Comment by olafgrabienski** [comment]
Sure, pleased to have another look at the sandbox site! Looks good overall, I also like the new language. I've found however two issues:

(1) The help text at content type setting is missing a word:
> Users who have the "View page-less content" will still be able to visit ...

Should be:
> Users who have the "View page-less content" **permission** will still be able to visit ...

(2) The "View page-less content" permission seems to not work anymore: I wasn't able to visit the full page of a "pageless" node as an Editor, got the "not found" error irrespective of the permission. (I had created an account for that role before, you should also be able to test it quickly when you give anonymous the permission temporarily.)

**Comment by alexfinnarn** [comment]
I made a small comment on the language https://github.com/backdrop/backdrop/pull/2286/files#r214403934. When I read the original first sentence, it made me not want to use the feature or think about how it could be used, because ain't nobody got time for a "Page not found" error. I think stating the purpose of the "pageless node" feature before the caveat makes it sound more powerful and less scary.

**Comment by herbdool** [comment]
Thanks @olafgrabienski I fixed the permission error and tweaked the help text (also based on other comments).

@kreynen so far none of the terms hit the spot. "pageless" isn't great either but to me it still seems better than any other suggestions. I think mentioning "path" in the help could improve the conveying the concept, though for a single term I wonder if "pathless content" works better. Though I'm not sure if it conveys as strong a concept as "pageless".

**Comment by kreynen** [comment]
But the content still has a path?  It's not pathless. This just prevents most users from accessing it using the path and would provide a flag for contrib modules like XML Sitemap and Linkit not to include nodes of these types.  

I know "beans" isn't a great name or acronym, but it does convey what makes these entities different and how they are designed to be used.  I understand the advantage of not introducing another fieldable thing that needs to be maintained into the core code base. Very smart. But if you don't want to give these a name like boxes or bricks, using terminology that accurately describes what is making this content type different from the others makes a lot more sense to me.

For this to work well, it's going to have to be embraced by contrib... that requires explaining it.

**Comment by herbdool** [comment]
@kreynen yeah it still has a path and a page so I agree the terminology isn't great. How about this for the permission: (I updated the PR with it) "view hidden path content" (or maybe it could be "view hidden path of content"). That's more accurate. It has a path but we've hidden it by default.

**Comment by herbdool** [comment]
I tweaked the permission again. Now it's "view hidden path display".

**Comment by herbdool** [comment]
Note that I'm trying to avoid labelling the content itself and hopefully avoid some bikeshedding. Maybe we can kick that can down the road in future PRs which try to add to the concept.

**Comment by herbdool** [comment]
Another quick review @olafgrabienski and anyone else?

**Comment by kreynen** [comment]
Are you going to change the node_type_get_types function in https://github.com/backdrop/backdrop/blob/1.x/core/modules/node/node.module#L380 so that every module won't have to get all the types and filter out the hidden/non-hidden options and add support to Views to be able to select all nodes of any content of either type?

**Comment by herbdool** [comment]
There was no discussion about altering `node_type_get_types()` but perhaps at minimum we can add a views filter to expose the hidden flag. I can't recall where else we've exposed config settings to views but seems like a good idea in this case.

**Comment by herbdool** [comment]
@kreynen since you expressed interest in the views filter would you be able to produce some sample code for me to add to the PR? I'm not sure if I have time to add it before feature freeze (currently only have small bits of time here and there and not likely to have much time on this Canadian long weekend.

**Comment by kreynen** [comment]
We have a long weekend in the US too and @kevincrafts (our UI/UX manager) is off next week, but I'll see what we can do.

**Comment by olafgrabienski** [comment]
@herbdool Did another quick sandbox review. Permission works now, nice! (After changing the permission, the setting applies however not immediately but only after clearing the "Page and else" cache.)

Regarding the terminology, I like 'path' instead of 'page' to avoid confusion with the Page content type. There is of course room for improvement, but I'd suggest to handle it in follow-up issues.

**Comment by alexfinnarn** [comment]
```php
<?php

/**
 * @file
 * Definition of views_handler_filter_node_hidden_path.
 */

/**
 * Filter by pageless access control.
 *
 * @ingroup views_filter_handlers
 */
class views_handler_filter_node_hidden_path extends views_handler_filter {
  function admin_summary() { }
  function operator_form(&$form, &$form_state) { }
  function can_expose() { return FALSE; }

  function query() {
    $table = $this->ensure_my_table();
    $node_types = config_get_names_with_prefix('node.type.');

    $enabled_types = array_filter($node_types, function($val) {
      return config_get($val, 'settings.hidden_path');
    });

    $enabled_types_names = array_map(function($val) {
      return str_replace('node.type.', '', $val);
    }, $enabled_types);
    
    $this->query->add_where_expression($this->options['group'], "$table.type IN (:types)", array(':types' => $enabled_types_names));
  }
}
```

By adding that views filter handler, an entry into `node_views_data()` and the autoload of the handler file, I was able to filter based on the hidden path setting.

The code could be cleaned up a little, and I took the example from `views_handler_filter_node_status.inc`.

**Comment by quicksketch** [comment]
I posted a short review to https://github.com/backdrop/backdrop/pull/2286#pullrequestreview-151562341. Code-wise, this looks good. 

I'm not sure I like `hidden path` any more than `pageless` but it is more technically accurate. I'd prefer that for internal storage even if we decide to change the user-facing label and description for the option later.

Can we lop the word "display" off the permission name and change it from "view hidden path display" to "view hidden paths" (similar to "view revisions")? The word "display" gets conflated with display modes in my mind.

**Comment by herbdool** [comment]
@quicksketch I've updated the permission wording.

@alexfinnarn thank you! I've gotten pretty close based on the class you provided. But I might be doing something wrong in `node_views_data`. I can get the filters and field to show up but getting some errors.

`Error: Call to undefined method views_handler_filter_node_hidden_path::pre_render() in view->render() (line 1273 of /app/core/modules/views/includes/view.inc).`

Here's my working copy in node_views_data():

```
// hidden path
  $data['node']['hidden_path'] = array(
    'title' => t('Hidden path'),
    'help' => t('Whether or not the content has a hidden path.'),
    'field' => array(
      'handler' => 'views_handler_filter_node_hidden_path',
      'click sortable' => TRUE,
      'output formats' => array(
        'visible-hidden' => array(t('Visible'), t('Hidden')),
      ),
    ),
    'filter' => array(
      'handler' => 'views_handler_filter_node_hidden_path',
      'label' => t('Paths are hidden'),
      'type' => 'yes-no',
      'use equal' => TRUE, // Use status = 1 instead of status <> 0 in WHERE statement
    ),
    'sort' => array(
      'handler' => 'views_handler_sort',
    ),
  );
```

Not sure if I can get this figured in time for code freeze. Hopefully but any help you can lend is much appreciated.

**Comment by herbdool** [comment]
@alexfinnarn I updated my previous comment. It works for me but the options don't show up in the filter (no yes/no option). Not clear yet on how to get it to appear.

**Comment by herbdool** [comment]
I'm wondering if at this point we can merge the PR as is and open a new issue for the view filter, since it doesn't block the main PR.

**Comment by herbdool** [comment]
Thanks @docwilmot I'll give it a try.

**Comment by herbdool** [comment]
Okay, ready for an other testing. I've now got a functioning views filter for including/excluding content with hidden paths with help from @alexfinnarn and @docwilmot.

cc @olafgrabienski

**Comment by olafgrabienski** [comment]
@herbdool Here my quick sandbox test results:

(1) From the help text at the content type:
> Users who have the "view hidden path" permission will still be able ...

Did you change the capitalized "**V**iew" to "**v**iew" on purpose? In my opinion, the capitalized version is more convenient.

(2) I wanted to add and configure the Views filter in the "Administer content" view and saw that it was already there and enabled ("Paths are hidden: Yes"), and it works! (I guess someone else had enabled the filter before. Can you double-check that it's not enabled by default?)

**Comment by herbdool** [comment]
@olafgrabienski 1) nope it wasn't on purpose. I can change it. 2) Actually I added it. I wanted to double-check on a different website.

**Comment by quicksketch** [comment]
One new issue with the views handler: https://github.com/backdrop/backdrop/pull/2286#pullrequestreview-151606175

My request from before was awkwardly made, so it didn't get conveyed quite:

> Can we lop the word "display" off the permission name and change it from "view hidden path display" to "view hidden paths" (similar to "view revisions")?

@herbdool the new permission name is now `view hidden path` (singular). Can you make that plural; `view hidden paths`? Other places like on the node type where the setting is just called `hidden_path` look good to me.

**Comment by herbdool** [comment]
@olafgrabienski one last review? The main difference is the slight change in permission wording and a fix to the tests.

**Comment by olafgrabienski** [comment]
@herbdool - done! Have tested the sandbox site regarding

- wording √
- hidden path setting vs. permission to view hidden path √
- Views filter √
- DBlog (no errors) √

So, RTBC in my opinion!

**Comment by quicksketch** [comment]
Thanks @herbdool and @olafgrabienski! This looks great! I've merged https://github.com/backdrop/backdrop/pull/2286 into 1.x for 1.11.0. Issue credit in the commit, as there are a lot of people to mention (https://github.com/backdrop/backdrop/commit/e48317db985e31ecd6d90738394b6a1ac43d6388).
